src/SAML2/XML/saml/Attribute.php

```diff
@@ -119,12 +106,15 @@ public function getNameFormat(): ?string
     /**
      * Set the value of the NameFormat-property
      *
-     * @param string|null $nameFormat
-     * @return void
+     * @param string|null $NameFormat
      */
-    public function setNameFormat(string $nameFormat = null): void
+    protected function setNameFormat(?string $NameFormat): void
     {
-        $this->NameFormat = $nameFormat;
+        if ($NameFormat === null) {
+            return;
```

Comment:
It may be useful to actually set it to `null`, so I would suggest to negate the if-statement and move the assertion in there..

Reply:
It defaults to `null`, so in case `null` is passed, `$this->NameFormat` was already `null`. Remember that the setters are intended only to set properties with all sanity checks needed during construction.

Reply:
But the class is not final, so it could potentially be called more than once from an extending class..
I know it's not very likely to happen, but I'd like to keep things consistent and I know we have a couple of cases where we do as I described..

Reply:
One of them:
https://github.com/simplesamlphp/saml2/pull/197/files#diff-5068cfce0ab6b548e1e6f9f7818e97afR158

Reply:
Uhm, that's a good point! That actually raises a bigger concern. We're removing checks in `toXML()` since the object is guaranteed to be properly created at all times. But if a class extends it, the properties are protected and can be accessed by the extending class, therefore breaking all guarantees for immutability.

I'm thinking we need `SAML2\XML\saml\Attribute` to be declared `final`. We can do that and keep `md\RequestedAttribute` the way it is by either moving all the functionality in `Attribute` into a trait, or by creating an abstract `SAML2\XML\saml\AbstractAttributeType` class, and make both `Attribute` and `RequestedAttribute` extend it.

What do you think?